src/util/readwritefile.cpp

```diff
@@ -5,46 +5,45 @@
 
 #include <util/readwritefile.h>
 
+#include <streams.h>
 #include <util/fs.h>
 
 #include <algorithm>
 #include <cstdio>
 #include <limits>
 #include <string>
 #include <utility>
+#include <vector>
 
-std::pair<bool,std::string> ReadBinaryFile(const fs::path &filename, size_t maxsize)
+template <typename T>
+std::optional<T> ReadBinaryFile(const fs::path& filename, size_t maxsize)
 {
-    FILE *f = fsbridge::fopen(filename, "rb");
-    if (f == nullptr)
-        return std::make_pair(false,"");
-    std::string retval;
-    char buffer[128];
-    do {
-        const size_t n = fread(buffer, 1, std::min(sizeof(buffer), maxsize - retval.size()), f);
-        // Check for reading errors so we don't return any data if we couldn't
-        // read the entire file (or up to maxsize)
-        if (ferror(f)) {
-            fclose(f);
-            return std::make_pair(false,"");
-        }
-        retval.append(buffer, buffer+n);
-    } while (!feof(f) && retval.size() < maxsize);
-    fclose(f);
-    return std::make_pair(true,retval);
+    std::FILE *f = fsbridge::fopen(filename, "rb");
+    if (f == nullptr) return {};
+    T output{};
+    size_t file_size = fs::file_size(filename);
+    output.resize(std::min(file_size, maxsize));
+    try {
+        AutoFile{f} >> Span{output};
+    } catch (const std::ios_base::failure&) {
+        return {};
+    }
+    return output;
 }
 
-bool WriteBinaryFile(const fs::path &filename, const std::string &data)
+template std::optional<std::string> ReadBinaryFile(const fs::path &filename, size_t maxsize);
+template std::optional<std::vector<unsigned char>> ReadBinaryFile(const fs::path &filename, size_t maxsize);
+
+template <typename T>
+bool WriteBinaryFile(const fs::path& filename, const T& data)
 {
-    FILE *f = fsbridge::fopen(filename, "wb");
-    if (f == nullptr)
-        return false;
-    if (fwrite(data.data(), 1, data.size(), f) != data.size()) {
-        fclose(f);
-        return false;
-    }
-    if (fclose(f) != 0) {
+    try {
+        AutoFile{fsbridge::fopen(filename, "wb")} << Span{data};
```

Comment:
Same as above, the [dumb](https://en.wikipedia.org/wiki/KISS_principle) version was easier to follow. The new one does not check whether `fclose(3)` failed, but it should. I think that is a serious deficiency in `AutoFile` itself.

`fwrite(3)` may succeed, but if a subsequent `fclose(3)` fails we should consider the data did not make it safely to disk and that the file is corrupted (`fclose(3)` writes any buffered data to disk using `fflush(3)`, so a failure at `fclose(3)` is as bad as failure at `fwrite(3)`).

Reply:
Logged as https://github.com/bitcoin/bitcoin/pull/29307

Reply:
@vasild I would also prefer to fix things in AutoFile, but perhaps add a few comments to explain what happens?

Reply:
This could happen:

1. fwrite() succeeds, returns ok the caller, some of the data is buffered in the OS and not yet on disk
2. fclose() is called, it tries to fflush() the buffered data to disk but fails due to IO error. The caller ignores the error returned by fclose()
3. the program continues with the wrong assumption that the data is safely on disk

Reply:
That's not good. I guess we also don't want to sync to disk for every field that's `>>`'d to a file though.

Reply:
Right. That's why I don't see a good solution. It is a design issue with `AutoFile` to flush/close from the destructor which can't signal the failure to the caller.